File: sdk/ruby/lib/debug_assist.rb
# frozen_string_literal: true

# DebugAssist SDK for Ruby
#
# Captures unhandled exceptions and reports them to DebugAssist.
#
# Usage:
#   require 'debug_assist'  # só isso
#
# Configuration (env vars):
#   DEBUG_ASSIST_API_KEY   — your API key
#   DEBUG_ASSIST_BASE_URL  — base URL (default: https://api.debug-assist.app)
#   DEBUG_ASSIST_PROJECT   — project name (default: unknown)
#   DEBUG_ASSIST_ENABLED=0 — disable SDK

require 'net/http'
require 'json'
require 'uri'

module DebugAssist
  API_KEY  = ENV.fetch('DEBUG_ASSIST_API_KEY', '').freeze
  BASE_URL = ENV.fetch('DEBUG_ASSIST_BASE_URL', 'https://api.debug-assist.app').chomp('/').freeze
  PROJECT  = ENV.fetch('DEBUG_ASSIST_PROJECT', 'unknown').freeze
  ENABLED  = ENV['DEBUG_ASSIST_ENABLED'] != '0'

  def self.send_diagnostic(message, exception_type, stack)
    return unless ENABLED && !API_KEY.empty?

    payload = JSON.generate(
      tipo: 'silent_backend_error',
      mensagem: message.to_s,
      contexto: {
        project_name:   PROJECT,
        exception_type: exception_type.to_s,
        stack:          stack.to_s
      }
    )

    uri  = URI("#{BASE_URL}/v1/diagnosticos")
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl     = uri.scheme == 'https'
    http.open_timeout = 10
    http.read_timeout = 10

    req = Net::HTTP::Post.new(uri.path.empty? ? '/' : uri.path)
    req['Content-Type']  = 'application/json'
    req['Authorization'] = "Bearer #{API_KEY}"
    req.body = payload

    http.request(req)
  rescue StandardError => e
    warn "[DebugAssist] Falha ao enviar diagnóstico: #{e.message}"
  end

  # Register at_exit hook
  if ENABLED && !API_KEY.empty?
    at_exit do
      err = $!
      next if err.nil? || err.is_a?(SystemExit)

      DebugAssist.send_diagnostic(
        err.message,
        err.class.name,
        err.backtrace&.join("\n") || ''
      )
    end
  end
end

File: sdk/csharp/DebugAssist.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;

namespace DebugAssistSDK
{
    /// <summary>
    /// DebugAssist SDK for C# — auto-captures unhandled exceptions and reports them to DebugAssist.
    ///
    /// Usage (one line in Program.cs):
    ///   DebugAssist.EnsureInitialized();
    ///
    /// Or with explicit key:
    ///   DebugAssist.Init("YOUR_API_KEY", "my-project");
    ///
    /// Configuration via env vars:
    ///   DEBUG_ASSIST_API_KEY   — your API key
    ///   DEBUG_ASSIST_BASE_URL  — base URL (default: https://debug-assist.onrender.com)
    ///   DEBUG_ASSIST_PROJECT   — project name (default: unknown)
    ///   DEBUG_ASSIST_ENABLED=0 — disable SDK
    /// </summary>
    public static class DebugAssist
    {
        private static string _apiKey      = Environment.GetEnvironmentVariable("DEBUG_ASSIST_API_KEY") ?? string.Empty;
        private static string _baseUrl     = (Environment.GetEnvironmentVariable("DEBUG_ASSIST_BASE_URL") ?? "https://debug-assist.onrender.com").TrimEnd('/');
        private static string _projectName = Environment.GetEnvironmentVariable("DEBUG_ASSIST_PROJECT") ?? "unknown";
        private static bool   _enabled     = Environment.GetEnvironmentVariable("DEBUG_ASSIST_ENABLED") != "0";
        private static bool   _initialized = false;
        private static readonly object _lock = new object();

        // Static HttpClient — reuse across calls (avoid socket exhaustion)
        private static readonly HttpClient _http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(10)
        };

        /// <summary>Triggers initialization from env vars. Place at the top of Program.cs.</summary>
        public static void EnsureInitialized() => RegisterHandlers();

        /// <summary>Initializes with an explicit API key.</summary>
        public static void Init(string apiKey, string projectName = "unknown")
        {
            lock (_lock)
            {
                _apiKey      = apiKey;
                _projectName = projectName;
            }
            RegisterHandlers();
        }

        private static void RegisterHandlers()
        {
            lock (_lock)
            {
                if (_initialized) return;
                if (!_enabled || string.IsNullOrWhiteSpace(_apiKey)) return;
                _initialized = true;
            }

            AppDomain.CurrentDomain.UnhandledException += (_, args) =>
            {
                if (args.ExceptionObject is Exception ex)
                    SendSync(ex.Message, ex.GetType().FullName ?? "Exception", ex.StackTrace ?? string.Empty);
            };

            TaskScheduler.UnobservedTaskException += (_, args) =>
            {
                args.SetObserved();
                var inner = args.Exception.InnerException ?? (Exception)args.Exception;
                SendSync(
                    inner.Message,
                    inner.GetType().FullName ?? "Exception",
                    inner.StackTrace ?? string.Empty
                );
            };
        }

        private static void SendSync(string message, string exceptionType, string stack)
        {
            try
            {
                var payload = JsonSerializer.Serialize(new
                {
                    tipo     = "silent_backend_error",
                    mensagem = message,
                    contexto = new
                    {
                        project_name    = _projectName,
                        exception_type  = exceptionType,
                        stack
                    }
                });

                using var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v1/diagnosticos")
                {
                    Content = new StringContent(payload, Encoding.UTF8, "application/json")
                };
                request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {_apiKey}");

                _http.Send(request);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"[DebugAssist] Falha ao enviar diagnóstico: {e.Message}");
            }
        }
    }
}
